imaging_tracking/admin.py

```diff
@@ -1,24 +1,129 @@
-from django.contrib.auth.models import User, Group
 from django.contrib.auth.admin import UserAdmin
+from django.contrib.auth.models import Group, User
+
+from experiments import auth
+from experiments.constants import Role
+from experiments.models.casbin import CasbinRule
+from experiments.models.measurement import Project
 from experiments.models.user import Profile
 
+from django.contrib.admin.widgets import FilteredSelectMultiple
+
 
 class ImagingTrackingAdminSite(AdminSite):
     site_header = 'Imaging Experiments Tracker'
 
 
+class CasbinForm(forms.ModelForm):
+    ptype = forms.CharField(widget=forms.HiddenInput(), initial='g')
+    user_id = forms.ChoiceField(label="User", choices=[])
+    project_id = forms.ChoiceField(label="Project", choices=[])
+    domain = forms.ChoiceField(label="Role", choices=[])
+
+    def __init__(self, obj=None, *args, **kwargs):
+        super(CasbinForm, self).__init__(obj, *args, **kwargs)
+        self.fields['user_id'].choices = [
+            (u.id, u.username) for u in User.objects.all().order_by('username')]
+        self.fields['domain'].choices = [(r.value, r.value) for r in Role]
+        self.fields['project_id'].choices = [
+            (p.id, p.name) for p in Project.objects.all().order_by('name')]
+
+    class Meta:
+        model = CasbinRule
+        exclude = ["action", "v3", "v4", "v5"]
+
+
+class CasbinRuleUserFilter(admin.SimpleListFilter):
+    """Filter project-roles for a user."""
+    title = 'User'
+    parameter_name = 'user_id'
+
+    def lookups(self, request, model_admin):
+        return [(str(u.id), u.username) for u in User.objects.all().order_by('username')]
+
+    def queryset(self, request, queryset):
+        value = self.value()
+        if value is None:
+            return queryset
+        return queryset.filter(ptype='g', user_id=value)
+
+
+class CasbinRuleProjectFilter(admin.SimpleListFilter):
+    """Filter project-roles for a project."""
+    title = 'Project'
+    parameter_name = 'project_id'
+
+    def lookups(self, request, model_admin):
+        return [(str(p.id), p.name) for p in Project.objects.all().order_by('name')]
+
+    def queryset(self, request, queryset):
+        value = self.value()
+        if value is None:
+            return queryset
+        return queryset.filter(ptype='g', project_id=value)
+
+
+class CasbinRuleRoleFilter(admin.SimpleListFilter):
+    """Filter project-roles for a role."""
+    title = 'Role'
+    parameter_name = 'domain'
+
+    def lookups(self, request, model_admin):
+        return [(r.value, r.value) for r in Role]
+
+    def queryset(self, request, queryset):
+        value = self.value()
+        if value is None:
+            return queryset
+        return queryset.filter(ptype='g', domain=value)
+
+
+class CasbinAdmin(admin.ModelAdmin):
+    model = CasbinRule
+    form = CasbinForm
+    actions = ["delete_role"]
+    list_display = ["user", "project", "get_role"]
+    list_display_links = None
```

Comment:
it's good to explain why we have this line

Reply:
Which one in particular?

Reply:
or maybe I misunderstood, but I still don't understand why `list_display_links = None`

Reply:
Prevents users from editing an existing ProjectRole, giving the users only the actions to delete or add a ProjectRole. 
To change an existing ProjectRole for a user you only add a new one that will override the previous one (logic implemented in `Authorization.add_role(...)`, which removes existing roles before adding the new one).

Reply:
Can you please explain this in a comment in the code?